src/Tomcat/TomcatController.ts

```diff
@@ -290,14 +319,22 @@ export class TomcatController {
         }
     }
 
-    private async deployPackage(serverInfo: TomcatServer, packagePath: string): Promise<void> {
-        const appName: string = path.basename(packagePath, path.extname(packagePath));
-        const appPath: string = path.join(serverInfo.getStoragePath(), 'webapps', appName);
+    private async deployWebapp(server: TomcatServer, webappPath: string): Promise<void> {
+        if (!server || !await fse.pathExists(webappPath)) {
+            return;
+        }
+        const appName: string = path.basename(webappPath, path.extname(webappPath));
+        const appPath: string = path.join(server.getStoragePath(), 'webapps', appName);
 
         await fse.remove(appPath);
         await fse.mkdirs(appPath);
-        Utility.trackTelemetryStep('deploy war');
-        await Utility.executeCMD(this._outputChannel, serverInfo.getName(), 'jar', { cwd: appPath }, 'xvf', `${packagePath}`);
+        if (webappPath.endsWith('.war')) {
```

Comment:
check with the webappPath validation first. If invalid we just return.

Then we calculate `appName` and other variables if it's valid

Reply:
need calculate appName first for appPath, so the calculation is inevitable